File: CopyRandomList.cc
#include <iostream>
using namespace std;

struct RandomListNode {
    int label;
    RandomListNode *next, *random;
    RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
};

//O(n) time and O(1) space
    RandomListNode *copyRandomList(RandomListNode *head) {

        RandomListNode* p = head;
        if (!head)
            return NULL;
        
        while (p) {
            RandomListNode* pnext = p->next;
            p->next = new RandomListNode(p->label);
            p->next->next = pnext;
            p = pnext;
        }
    
        RandomListNode* newHead = head->next;
        for (p = head; p ; )
        {
           p->next->random = p->random ? p->random->next : NULL;
           p = p->next->next;
        }
    
        for (p = head; p; )
        {
            RandomListNode* pnext = p->next->next;
            p->next->next = pnext ? pnext->next : NULL;
            p->next = pnext;
            p = pnext;
        }
        
        return newHead;
    
    }

//Another solution is using a hashmap to store the mapping from original nodes
//to their copies. It takes two passes and O(n) extra space 


int main(int argc, char** argv)
{
    RandomListNode head(-1);
    RandomListNode* newHead = copyRandomList(&head);
    RandomListNode* p = newHead;
    while(p) {
        string randomLabel = p->random ? to_string(p->random->label) : "NULL";
        cout << p->label << ", " << randomLabel << endl;
        p = p->next;
    }
    return 0;
}

File: CopyRandomList.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class CopyRandomList {

	// O(n) time and O(n) space
    public RandomListNode copyRandomList0(RandomListNode head) {
        RandomListNode newDummy = new RandomListNode(-1);
        RandomListNode pre = newDummy;
        Map<RandomListNode, RandomListNode> nodeMap = new HashMap<>();
        for (RandomListNode cur = head; cur != null; cur = cur.next) {
            RandomListNode newCur = new RandomListNode(cur.label);
            nodeMap.put(cur, newCur);
            pre.next = newCur;
            pre = newCur;
        }
        RandomListNode cur = head;
        for (RandomListNode newCur = newDummy.next; newCur != null; newCur = newCur.next, cur = cur.next) {
            newCur.random = nodeMap.get(cur.random);
        }
        return newDummy.next;
    }
    
    // O(n) time and O(1) space
    public RandomListNode copyRandomList(RandomListNode head) {
        for (RandomListNode cur = head; cur != null; cur = cur.next.next) {
            RandomListNode curCopy = new RandomListNode(cur.label);
            curCopy.next = cur.next;
            cur.next = curCopy;
        }
        for (RandomListNode cur = head; cur != null; cur = cur.next.next) {
            cur.next.random = (cur.random == null) ? null : cur.random.next;
        }
        RandomListNode newDummy = new RandomListNode(-1);
        RandomListNode pre = newDummy;
        for (RandomListNode cur = head; cur != null; cur = cur.next) {
            pre.next = cur.next;
            cur.next = cur.next.next;
            pre = pre.next;
        }
        return newDummy.next;
    }
}